server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -3927,9 +3926,11 @@ public void testScheduledRefresh() throws Exception {
         assertTrue(primary.searchIdleTime() >= TimeValue.ZERO.millis());
         primary.flushOnIdle(0);
         logger.info("--> scheduledRefresh(future5)");
```

Comment:
I have some questions related to the test (and not specifically to this line) that I don't understand:
* Why is `assertFalse(future4.actionGet());`? The comment before future4 says `while shard is search active and ensure scheduleRefresh(...) makes documen visible:`. So I am not sure why the scheduled refresh returns false? I think that the shard was search idle, and that's why the scheduled refresh is false. Right? We should probably correct that comment.
* Is the `primary.flushOnIdle(0);` actually flushing the shard? Doesn't flush also refresh? So how come `assertTrue(future5.actionGet());`? I would expect that since the shard was refreshed during the flush, that the scheduled refresh of future5 would return false.

Until I understand the test, I'm not sure I can follow the proposed solution. Any answers would be very helpful to me!

Reply:
>  I think that the shard was search idle, and that's why the scheduled refresh is false. Right? We should probably correct that comment.

Yes, that's the reason why it returns false in this case, we should update the comment 👍.

>  Doesn't flush also refresh? 

Notice that the refresh executed after a flush concerns the internal searcher (https://github.com/elastic/elasticsearch/blob/6ae652f90e7b15d9eef81fa745da299d39845655/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L2225) whereas the scheduled refresh concerns the external searcher (https://github.com/elastic/elasticsearch/blob/6ae652f90e7b15d9eef81fa745da299d39845655/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L2020) hence the need for the explicit refresh.

If it's the flush thread racing against the scheduled refresh call what's causing this test failure, I'm not sure if the proposed solution will fix it, since it will return false always as the refresh will be a no-op.

Reply:
Thanks @fcofdez for the clarification, I see!

> If it's the flush thread racing against the scheduled refresh call what's causing this test failure, I'm not sure if the proposed solution will fix it, since it will return false always as the refresh will be a no-op.

Why is that? I think it actually may work. If there's a race with flush and refresh, the refresh may return false due to the race. But once the flush is executed, it makes `active` flag to be `false`, which should make the `scheduledRefresh()` finally refresh. So an assertBusy with scheduledRefresh should finally succeed. Right?

Reply:
>  which should make the scheduledRefresh() finally refresh. So an assertBusy with scheduledRefresh should finally succeed. Right?

But if the refresh results in a no-op (there're no pending docs in the Lucene IndexWriter) the refresh will return false  , right?

Reply:
But for some reason there's another thread calling refresh and that's what's causing the test to fail, either the scheduled refresh is not able to acquire the reader manager lock or the scheduled refresh results in a no-op, in both cases I think that the assertBusy won't help with this. I would like for us to reproduce this issue with some sleeps or something.

Reply:
Oh so far I thought we concluded that there's a race between the flush and the refresh. I didn't understand there was another refresh request going on. For my sake, where did we confirm/see this other refresh request (e.g., in the logs)?

Reply:
Looking more closely into the logs, it must be a concurrent refresh what's causing this, because the flush is executed _after_ the test fails:
```
[2024-03-21T12:21:28,090][INFO ][o.e.i.s.IndexShardTests  ] [testScheduledRefresh] --> scheduledRefresh(future5) |  
-- | --
  | [2024-03-21T12:21:28,090][TRACE][o.e.i.s.IndexShard       ] [testScheduledRefresh] [test][0] scheduledRefresh: refresh with source [schedule] |  
  | [2024-03-21T12:21:28,091][INFO ][o.e.i.s.IndexShardTests  ] [testScheduledRefresh] after test |  
  | [2024-03-21T12:21:28,093][TRACE][o.e.i.s.IndexShard       ] [org.elasticsearch.index.shard.IndexShardTests] [test][0] flush with FlushRequest{waitIfOngoing=false, force=false}
```
I think that @arteam should dig deeper into what's going on here.

Reply:
@fcofdez After looking at the logs you mentioned (thanks!), I believe it's not another refresh. But it's actually exactly because the flush may happen after the future5's refresh request! If the flush happens after the scheduledRefresh call, then the scheduledRefresh (similarly to future4) is not executed as it's still search idle (`active` is true). The flush is the operation that turns `active` to false, and is able to make the scheduledRefresh actually execute a refresh.

Thus, I think finally that the best solution is to do the following:

```
        long flushesBefore = primary.flushStats().getTotal();
        primary.flushOnIdle(0);
        assertBusy(() -> {
            assertThat(primary.flushStats().getTotal(), greaterThan(flushesBefore));
        });
```

Reply:
but the log line ` [2024-03-21T12:21:28,090][TRACE][o.e.i.s.IndexShard       ] [testScheduledRefresh] [test][0] scheduledRefresh: refresh with source [schedule]` hints that the scheduled refresh goes through

Reply:
I see. Hmm, this is still a bit messy to the asyncronicity of the flush and the refresh. I believe this is still the issue rather than a second refresh. 

Here's one possibility I see that this could error:
- `flushOnIdle` sets `active` to false, but not yet schedules the flush.
- `scheduledRefresh` can go into `maybeRefresh` because it sees `active` as false.
- So we can now have two racing `InternalEngine#refresh()` calls (one with internal scope due to the flush, and one for external due to the scheduled refresh).
- At the same point, they do `if (store.tryIncRef()) {`, and the flush wins while the refresh does not win.

But indeed it seems worth @arteam to figure it out and confirm, e.g., by reproducing the issue somehow. To also ensure we don't have other refreshes.

Reply:
Let me try to see if I manage something similar to `SearchIdleIT#ensureNoPendingScheduledRefresh`

Reply:
@kingherc I've added a check that the flush finishes before the refresh and a check that there are no pending refreshes.

Reply:
> At the same point, they do if (store.tryIncRef()) {, and the flush wins while the refresh does not win.

@kingherc but the store shouldn't be closed while the test is running 🤔 (both flush and refresh should be able to incRef the store ref unless the shard is closing)

Reply:
The only possibility is that https://github.com/elastic/elasticsearch/blob/bdc65cbc901ae059ef67da9a76cac870fd8c1ea1/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L2046 returns `false` due to a concurrent refresh making this one a no-op.

Reply:
@fcofdez , when I was quickly checking the code, I mistakenly thought `tryIncRef` does not have while-loop internally to retry, but it does, so my thinking was erroneous. Needs more in-depth investigation than my quick going-back-and-forth in the code finally :D 

Still I believe it was not the intention of the test to let both the flush & the refresh to race with each other. However, I cannot say for sure why the racing results in refresh returning false. It would be good to understand why @arteam , plus whether there's some other refresh going on as @fcofdez suggests. @arteam were you able to reproduce it? Maybe we can add some more tracing to see if another refresh is going on.